# Fuzzy logic exercises

In this notebook, you will write code to compute the cardinality of finite fuzzy sets and basic fuzzy logic opreators: union, intersection, and complement. For this exercise, you should not use any libraries, but rather use only basic Python.

Let's define some fuzzy sets:

In [24]:
# Universal set
U = ('a', 'b', 'c', 'd', 'e', 'f', 'g')

# A and B are finite fuzzy sets:
A = { 'a': 0.4, 'b': 1, 'c': 0.6, 'd': 0.0, 'e' : 0.5, 'f' : 0.3, 'g' : 0.9  }
B = { 'a': 0.6, 'b': 0.2, 'c': 0.6, 'd': 0.7, 'e' : 0.99, 'f' : 0.5, 'g' : 0  }

## Cardinality and complement

Implement code that can outputs the cardinality of $A$ and the cardinality of $B$:

In [25]:
# implement your code to compute the cardinality of A and B here:
print(sum(A.values()))
print(sum(B.values()))

3.6999999999999997
3.59


The cardinality of $A$ should be 3.7

The cardinality of $B$ should be 3.59

Now, implement code that computes C as the complement of $A$, that is $C = \bar{A}$:

In [26]:
# implement your code to compute the complement of A here:
C = dict()
for (key, val) in A.items():
    C[key] = 1 - val
C

{'a': 0.6,
 'b': 0,
 'c': 0.4,
 'd': 1.0,
 'e': 0.5,
 'f': 0.7,
 'g': 0.09999999999999998}

## Union and intersection

Implement code that computes $I$ as the intersection of $A$ and $B$, that is $I = A \cap B$:

In [27]:
# implement your code to compute the intersection between A and B here:
D = dict()
for (keyA, valueA) in A.items():
    for (keyB, valueB) in B.items():
        if(keyA == keyB):
            D[keyA] = min(valueA,valueB)
print(D)

{'a': 0.4, 'b': 0.2, 'c': 0.6, 'd': 0.0, 'e': 0.5, 'f': 0.3, 'g': 0}


Finally, write code that computes $N$ as the union of $A$ and $B$, that is $N = A \cup B$

In [28]:
# implement your code to compute the union of A and B here:
E = dict()
for (keyA, valueA) in A.items():
    for (keyB, valueB) in B.items():
        if(keyA == keyB):
            E[keyA] = max(valueA,valueB)
print(E)

{'a': 0.6, 'b': 1, 'c': 0.6, 'd': 0.7, 'e': 0.99, 'f': 0.5, 'g': 0.9}


Great! You reached the end of the exercises in this notebook. However, note that the membership degree was specified for all elements of $U$ in both the dicts ``A`` and ``B``, even for elements with member degree 0, for instance ``d``in ``A`` (``A = { ... 'd': 0.0, ...  }``. If you haven't already done so, try to make your code robust to fuzzy set declarations as dicts that only contain a key-pairs for the elements with a positive membership degree. Hence, make the code work even if you remove ``d`` from the ``A`` and ``g`` from ``B`` in the declaration of the dicts at the top of this notebook.  
